packages/fxa-content-server/app/scripts/models/account.js

```diff
@@ -372,6 +382,128 @@ const Account = Backbone.Model.extend(
+      try {
+        let ecosystemAnonId = this.get('ecosystemAnonId');
+        if (ecosystemAnonId) {
+          return;
+        }
+
+        // If no data pipeline keys are available, we can't generate the ecosystemAnonId, abort
+        if (this._ecosystemAnonIdPublicKeys.length <= 0) {
+          return;
+        }
+
+        // Pick a random public key
+        const randomEcosystemAnonIdPublicKey = this._ecosystemAnonIdPublicKeys[
+          Math.floor(Math.random() * this._ecosystemAnonIdPublicKeys.length)
+        ];
+
+        //
+        if (options.kB) {
+          ecosystemAnonId = await ecosystemAnonKeys.generateEcosystemAnonID(
+            options.kB,
+            randomEcosystemAnonIdPublicKey
+          );
+        } else if (this.canFetchKeys()) {
+          const keys = await this.accountKeys();
+          ecosystemAnonId = await ecosystemAnonKeys.generateEcosystemAnonID(
+            keys.kB,
+            randomEcosystemAnonIdPublicKey
+          );
+        } else if (options.password) {
+          // TODO: Using sessionReauth to get keys could potentially pose some issues if the user
+          // is attempting to come from an unverified session. FxA only allows fetching keys from
+          // sessions that are verified (ex. performed some email verification loop).
+          const res = await this._fxaClient.sessionReauth(
+            this.get('sessionToken'),
+            this.get('email'),
+            options.password,
+            {
+              wantsKeys: () => true, // Ensure keys are requested
+              has: () => false,
+              isSync: () => false,
+            },
+            {
+              keys: true,
+
+              // TODO: The verification reason needs to be updated to be specific for ecosystemAnonId
+              // There are checks in the auth-server to send emails depending on the `reason` value specified.
+              reason: VerificationReasons.RECOVERY_KEY,
+            }
+          );
+
+          if (res.keyFetchToken && res.unwrapBKey) {
+            const keys = await this._fxaClient.accountKeys(
+              res.keyFetchToken,
+              res.unwrapBKey
+            );
+            ecosystemAnonId = await ecosystemAnonKeys.generateEcosystemAnonID(
+              keys.kB,
+              randomEcosystemAnonIdPublicKey
+            );
+          }
+        }
+
+        // Update localstorage and report new ecosystemAnonId to auth-server
+        if (ecosystemAnonId) {
+          this.set('ecosystemAnonId', ecosystemAnonId);
+
+          await this._fxaClient.updateEcosystemAnonId(
+            this.get('sessionToken'),
+
+            // TODO: We need to ensure that the ecosystemAnonId is stored and updated in correct format.
+            // ie. hex or base64
+            ecosystemAnonId
+          );
+        }
+      } catch (err) {
+        // In the event of errors, we don't want to block any user actions, just log it
```

Comment:
If we get a 412 error, then we tried to update the anon_id, but another client beat us to it. We should handle that error by clearing the newly-created value out of localstorage and refetching the new value from the server. OK to handle this in a followup bug.

Reply:
In general, I think if we get a 4xx or 5xx error here, we probably need to discard the new ecosystemAnonId value until we can phone home successfully.

Reply:
Filed new issue here: https://github.com/mozilla/fxa/issues/6017